### PR TITLE
Add LatteSwap fetch TVL

### PR DESCRIPTION
**Pull request by cappuccino-kub**
##### Twitter Link:  
- https://twitter.com/latteswapOG

##### List of audit links if any:
- https://app.inspex.co/library/latteswap#?scope=latteswap-farm-swap-token-and-vault
- https://github.com/latteswap-official/latteswap-contract/blob/main/audit/protocol/PeckShield-Audit-Report-LatteSwap-v1.0.pdf

##### Website Link:
- https://app.latteswap.com

##### Logo (High resolution, preferably in .svg and .png, for application on both white and black backgrounds. Will be shown with rounded borders):
- https://drive.google.com/file/d/1pggcIqcQjA5u04bBOBIKoIePMHgLV4fv/view

##### Current TVL:
- $874,245,943.14 USD

##### Chain:
- BSC

##### Coingecko ID (so your TVL can appear on Coingecko): (https://api.coingecko.com/api/v3/coins/list)
--

##### Coinmarketcap ID (so your TVL can appear on Coinmarketcap): (https://api.coinmarketcap.com/data-api/v3/map/all?listing_status=active,inactive,untracked&start=1&limit=10000)
--

##### Short Description (to be shown on DefiLlama):
- Decentralized exchange with integrated NFT functionalities operating on the Binance Smart Chain (BSC). It is a one-stop-shop for traders, yield farmers, and NFT collectors across the Blockchain ecosystem.

##### Token address and ticker if any:
- 0x8d78c2ff1fb4fba08c7691dfeac7bb425a91c81a

##### Category (Yield/Dexes/Lending/Minting/Assets/Insurance/Options/Indexes/Staking) *Please choose only one:
- Dexes

##### Oracle used (Chainlink/Band/API3/TWAP or any other that you are using):
--

##### forkedFrom (Does your project originate from another project):
- SushiSwapV2

##### methodology (what is being counted as tvl, how is tvl being calculated):
- 'TVL accounts for the liquidity on all AMM pools and single asset staking pools, using the same source as TVL on https://app.latteswap.com/yield page.

### CODE REVIEW

**Comment by waynebruce0x** [comment]
Can this be calculated using an on-chain adapter?

**Comment by elgrayfrappuccino** [comment]
> Can this be calculated using an on-chain adapter?

Hi @waynebruce0x . To my understanding, you are asking if we could write a contract to calculate the TVL?. In that case, it might be possible, however I would not suggest doing so since our TVL came from 3 major sources

1. PancakeSwap LP (LATTE-BUSD) pool: This one is quite straightforward, you could write a contract to read the pool reserve and convert them to USD. 

2. LatteSwap LP pair pools: This is the same as above except the contract address must be LatteSwap's contract instead of PancakeSwap. However, since we are making more partnerships and let them listed on our site, the number of pools will keep growing so it would be hard to keep track of the newer pools using any on-chain solutions. For now, we are using The Graph to keep track of all pools. (which I believe that most of AMM are using)

3. Single asset staking pools: Its issue is the same as the number 2. Plus, we are having some pools that might need more logic to calculate the TVL (for example ibAlpaca token which needs to convert back to Alpaca before calculate the price) so it would be even harder to write an on-chain solution for that.

In conclusion, the TVL might be able to calculate using a contract/adaptor but I would not suggest to keep track of any DEXs' TVL using any on-chain solutions unless you could make sure it will be updated when the new pools are added and it can have a special calculation logic for some specific pools.

Does that answer you question?

**Comment by 0xngmi** [comment]
- LatteSwap LP pair pools: These are not a big issue for us, we can just get them from the factory and price them
- Single asset staking pools: Could you create a separate endpoint with the tvl of these?

**Comment by mint-chocolate-frappuccino** [comment]
Can we instead give you an open-source version of our api and let you validate our TVL domain logic?

**Comment by mint-chocolate-frappuccino** [comment]
Please note that we use this approach because of reusability,  we can provide a codebase for you for transparency.

**Comment by 0xngmi** [comment]
> Can we instead give you an open-source version of our api and let you validate our TVL domain logic?

i'd prefer other methods since then we can get historical values but this works too. Only hard requirement from our side is that you separate the tvl of each thing

**Comment by mint-chocolate-frappuccino** [comment]
okay, we will open source our API for TVL calculation and share you guys in a PR description, sounds good?

**Comment by cappuccino-kub** [comment]
Here is the source of our API for TVL calculation
https://github.com/latteswap-official/latteswap-open-api/blob/master/src/usecases/amm/get_total_value_locked.ts

**Comment by 0xngmi** [comment]
Could you make the API return the value locked on the LATTE pool separately?
image

For now it's very small, but if it grows large in the future it would represent a deviation from our standard methodology

**Comment by cappuccino-kub** [comment]
We update API for fetch TVL and update code for this PR
This is source code about the API https://github.com/latteswap-official/latteswap-open-api/tree/master/src/usecases/amm/defi_llama

**Comment by 0xngmi** [comment]
Thanks for making the API changes, given that your PR changed 10k lines of package-locked, I opted for applying the changes directly instead of merging. I also separated LATTE staking from the other sources of tvl, in order to follow the same methodology we use for our other projects (eg: pancakeswap, aave..)

**Comment by elgrayfrappuccino** [comment]
@0xngmi So the code was pushed to main directly and we could close this PR, correct?
